controller/authorize_blackbox_test.go

```diff
@@ -0,0 +1,301 @@
+
+	ctx := context.Background()
+	goaCtx := goa.NewContext(goa.WithAction(ctx, "AuthorizeTest"), rw, req, prms)
+	authorizeCtx, err := app.NewAuthorizeAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.Nil(t, err)
+
+	err = ctrl.Authorize(authorizeCtx)
+	require.Nil(t, err)
+
+	require.Equal(t, 307, rw.Code) // redirect to keycloak login page.
+
+	locationString := rw.HeaderMap["Location"][0]
+	locationUrl, err := url.Parse(locationString)
+	require.Nil(t, err)
+
+	allQueryParameters := locationUrl.Query()
+
+	require.NotNil(t, allQueryParameters)
+	require.NotNil(t, allQueryParameters["state"][0])
+
+	returnedState := allQueryParameters["state"][0]
+
+	u = &url.URL{
+		Path: fmt.Sprintf(client.CallbackAuthorizePath()),
+	}
+
+	prms = url.Values{
+		"state": {returnedState},
+		"code":  {"SOME_OAUTH2.0_CODE"},
+	}
+
+	ctx = context.Background()
+	rw = httptest.NewRecorder()
+
+	req, err = http.NewRequest("GET", u.String(), nil)
+	require.Nil(t, err)
+
+	// The OAuth code is sent as a query parameter by calling /api/login?code=_SOME_CODE_&state=_SOME_STATE_
+	// The request originates from Keycloak after a valid authorization by the end user.
+	refererKeycloakUrl := "https://keycloak-url.example.org/path-of-login"
+	req.Header.Add("referer", refererKeycloakUrl)
+
+	goaCtx = goa.NewContext(goa.WithAction(ctx, "AuthorizecallbackTest"), rw, req, prms)
+	callbackCtx, err := app.NewCallbackAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.Nil(t, err)
+
+	err = ctrl.Callback(callbackCtx)
+	require.Nil(t, err)
+
+	require.Equal(t, 307, rw.Code) // redirect to keycloak login page.
+
+	require.Contains(t, rw.Header().Get("Location"), redirectURI)
+
+	locationString = rw.HeaderMap["Location"][0]
+	locationUrl, err = url.Parse(locationString)
+	require.Nil(t, err)
+
+	allQueryParameters = locationUrl.Query()
+
+	require.NotNil(t, allQueryParameters)
+	require.NotNil(t, allQueryParameters["state"][0])
+	require.NotNil(t, allQueryParameters["code"][0])
+	require.Equal(t, returnedState, allQueryParameters["state"][0])
+}
+
+func (rest *TestAuthorizeREST) TestAuthorizeCallbackBadRequest() {
+	t := rest.T()
+
+	rw := httptest.NewRecorder()
+	u := &url.URL{
+		Path: fmt.Sprintf(client.CallbackAuthorizePath()),
+	}
+	req, err := http.NewRequest("GET", u.String(), nil)
+	require.Nil(t, err)
+
+	prms := url.Values{}
+	prms.Add("code", "SOME_OAUTH2.0_CODE")
+	prms.Add("state", uuid.NewV4().String())
+
+	ctx := context.Background()
+
+	prms.Del("code")
+	goaCtx := goa.NewContext(goa.WithAction(ctx, "AuthorizeTest"), rw, req, prms)
+	callbackCtx, err := app.NewAuthorizeAuthorizeContext(goaCtx, req, goa.New("LoginService"))
+	require.NotNil(t, err)
+	require.Contains(t, err.Error(), "400 invalid_request: missing required parameter")
+	jsonapi.JSONErrorResponse(callbackCtx, err)
+	require.Equal(t, 400, rw.Code)
+
+	prms.Del("state")
+	goaCtx = goa.NewContext(goa.WithAction(ctx, "AuthorizeTest"), rw, req, prms)
```

Comment:
Add code param to make sure it fails because of missing state. Not code. And extract this code actually in a separate function and reuse for both params.